docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,68 @@
+The code generator is free to ignore the `HoistedLocal` modifier if it can prove that this makes no observable difference in the execution of the generated program. This can be observable in diagnostics since it may mean the value of a local with the `HoistedLocal` modifier will not be available after certain suspension points.
+
+Async methods have some temporary restrictions with may be lifted later:
+* The `tail` prefix is forbidden
+* Usage of the `localloc` instruction is forbidden
+* Pinning locals may not be marked `HoistedLocal`
+
+Other restrictions are likely to be permanent, including
+* By-ref locals cannot be hoisted across suspension points
+* Suspension points may not appear in exception handling blocks.
+* Only four types will be supported as the return type for "runtime-async" methods: `System.Threading.Task`, `System.Threading.ValueTask`, `System.Threading.Task<T>`, and `System.Threading.ValueTask<T>`
+
+All async methods effectively have two entry points, or signatures. The first signature is the one present in the above code: a `Task` or `ValueTask` returning method. The second signature is a "secondary signature", described in further detail in [I.8.6.1.5 Method signatures].
+
+Callers may retrieve a Task-equivalent return type from an async method via calling the "Task-equivalent signature". This functionality is available in both sync and async methods.
+
+### I.8.6.1.5 Method signatures
+
+The list of relevant components is augmented to include sync vs. async method types. Async methods have some additions to normal signature compatibility.
+
+Async methods are capable of calling certain methods using an "async signature." The target method must return one of `Task`, `ValueTask`, `Task<T>`, or `ValueTask<T>`.
+
+The async signature is generated based on the primary signature. The transformation is as follows:
+* If the target method return type is `Task` or `ValueTask`, the return type of the secondary signature is `void` with the first custom modifier on the return type being the original type.
+* Otherwise, the return type is the type argument of the return type (either ``Task`1`` or ``ValueTask`1``) with the first custom modifier on the return type being the original type.
```

Comment:
> with the first custom modifier on the return type being the original type

Consider generic methods such as the invoke method of `Func<Task>`:
```il
instance !0 class [System.Runtime]System.Func`1<class [System.Runtime]System.Threading.Tasks.Task>::Invoke()
```

Which of these would the runtime async signature be?

```il
// Use the type argument
instance modreq(!0) void class [System.Runtime]System.Func`1<class [System.Runtime]System.Threading.Tasks.Task>::Invoke()
```

```il
// Use Task directly
instance modreq([System.Runtime]System.ThreadingTasks.Task) void class [System.Runtime]System.Func`1<class [System.Runtime]System.Threading.Tasks.Task>::Invoke()
```

Reply:
One rule that we had in the prototype is that methods do not become async via generic substitution. 

Basically given `T Foo<T>()`, the only thing you can get from `Foo<Task>()` is to call it and get a `Task` result. You can't call it via modreq'd signature. 

Most likely we should keep the rule. Would it make this case resolved?

Reply:
> Would it make this case resolved?

It would, but I'm concerned about knock-on effects. Would we start seeing users defining `AsyncFunc` delegate types to avoid intermediate `Task` allocations and recover perf?

Reply:
In particular, `ASP.NET` is going to use a lot of `Task`-returning `Func`s. Are they in a performance pit because of this?

Reply:
See https://github.com/dotnet/runtime/issues/109632